mutable_buffer/src/table.rs

```diff
@@ -824,55 +814,75 @@ impl Table {
             })
     }
 
-    /// Returns (name, index) pairs for all named columns
-    fn column_names_with_index<'a>(
-        &self,
-        chunk: &Chunk,
-        columns: &[&'a str],
-    ) -> Result<Vec<(&'a str, usize)>> {
-        columns
-            .iter()
-            .map(|&column_name| {
-                let column_index = self.column_index(chunk, column_name)?;
-
-                Ok((column_name, column_index))
-            })
-            .collect()
-    }
+    /// Returns the column selection for all the columns in this table, orderd
+    /// by table name
+    fn all_columns_selection<'a>(&self, chunk: &'a Chunk) -> Result<TableColSelection<'a>> {
+        let selection = TableColSelection::with_capacity(self.column_id_to_index.len());
 
-    /// Convert all columns to an arrow record batch
-    pub fn all_to_arrow(&self, chunk: &Chunk) -> Result<RecordBatch> {
-        let mut requested_columns_with_index = self
-            .column_id_to_index
-            .iter()
-            .map(|(&column_id, &column_index)| {
+        let selection = self.column_id_to_index.iter().try_fold(
+            selection,
+            |selection, (&column_id, &column_index)| {
                 let column_name = chunk.dictionary.lookup_id(column_id).context(
                     ColumnIdNotFoundInDictionary {
                         column_id,
                         chunk: chunk.id,
                     },
                 )?;
-                Ok((column_name, column_index))
-            })
-            .collect::<Result<Vec<_>>>()?;
+                Ok(selection.add_column(column_name, column_index))
+            },
+        )?;
 
-        requested_columns_with_index.sort_by(|(a, _), (b, _)| a.cmp(b));
+        // sort so the columns always come out in a predictable name
+        Ok(selection.sort_by_name())
+    }
+
+    /// Returns a column selection for just the specified columns
+    fn specific_columns_selection<'a>(
+        &self,
+        chunk: &'a Chunk,
+        columns: &'a [&'a str],
+    ) -> Result<TableColSelection<'a>> {
+        let selection = TableColSelection::with_capacity(columns.len());
+        columns
+            .iter()
+            .try_fold(selection, |selection, &column_name| {
+                let column_index = self.column_index(chunk, column_name)?;
+
+                Ok(selection.add_column(column_name, column_index))
+            })
+    }
 
-        self.to_arrow_impl(chunk, &requested_columns_with_index)
+    /// Converts this table to an arrow record batch.
+    ///
+    /// If requested_columns is empty (`[]`), retrieve all columns in
+    /// the table
+    ///
+    /// TODO: make this take a ColumnSelection structure rather than &[&str]
+    /// directly.
```

Comment:
```suggestion
    ///
    /// If requested_columns is empty (`[]`), retrieve all columns in
    /// the table
    ///
    /// TODO: make this take a ColumnSelection structure rather than &[&str]
    /// directly.
```

LOL I finished the TODO in this PR